Feature/text annotation "Label"

canavandl opened the pull request:
Added Label annotation class.

In future, allows for refactoring ofchart title, axis label, (etc.) code to use Label model.

UX question: the Label doesn't resize on zoom events. Should it?

canavandl commented:
Should add bounding box for text (transparent by default).

canavandl commented:
Note: angle props don't seem to be working

bryevdv commented:
@canavandl can you more specific? For instance, they seem to be working here:

http://bokeh.pydata.org/en/latest/docs/gallery/burtin.html

bryevdv commented on sphinx/source/docs/user_guide/plotting.rst:
```diff
@@ -391,6 +391,17 @@ left or top) default will extend the annotation to the edge of the plot area.
 .. bokeh-plot:: source/docs/user_guide/source_examples/plotting_box_annotation.py
     :source-position: above
 
+<<<<<<< HEAD
```
looks like some merge conflict junk snuck in

bryevdv commented on sphinx/source/docs/user_guide/plotting.rst:
```diff
@@ -391,6 +391,17 @@ left or top) default will extend the annotation to the edge of the plot area.
+using the Label annotation.
+
+.. bokeh-plot:: source/docs/user_guide/source_examples/plotting_label.py
+=======
```
merge junk

bryevdv commented on sphinx/source/docs/user_guide/plotting.rst:
```diff
@@ -400,6 +411,7 @@ Spans (line-type annotations) have a single dimension (width or height) and
 extend to the edge of the plot area.
 
 .. bokeh-plot:: source/docs/user_guide/source_examples/plotting_span.py
+>>>>>>> master
```
merge junk

bryevdv commented on sphinx/source/docs/user_guide/source_examples/plotting_label.py:
```diff
@@ -0,0 +1,29 @@
+versicolor_label = Label(text='Versicolor', x=4.5, y=1.2, **label_props)
+virginica_label = Label(text='Virginica', x=5.5, y=2.1, **label_props)
+
+p.renderers.extend([versicolor_label, virginica_label])
```
Some suggested visual changes:
```
p.circle(flowers["petal_length"], flowers["petal_width"],
        color=flowers["color"], fill_alpha=0.3, size=10)
p.line(x=[3, 7], y=[2.25, 1.05], color='orange', alpha=0.5, line_dash='8 4', line_width=3)

label_props = dict(label_text_font_size='40pt',
                   label_text_align='center',
                   label_text_alpha=0.5)

versicolor_label = Label(text='Versicolor', x=5.3, y=1.1, label_text_color="green", **label_props)
virginica_label = Label(text='Virginica', x=4.4, y=2.2, label_text_color="blue", **label_props)
```

bryevdv replied:
<img width="589" alt="screen shot 2015-10-21 at 12 26 44 am" src="https://cloud.githubusercontent.com/assets/1078448/10629246/74e06a2a-778a-11e5-9c40-cedf81247ee2.png">

canavandl commented:
Removed merge cruft and updated user guide examples to @bryevdv 's styling. Should be ready to merge

bryevdv commented:
Is the plan to do linked/glyph annotations in a separate PR?

canavandl commented:
<img width="615" alt="screen shot 2015-11-02 at 12 37 29 pm" src="https://cloud.githubusercontent.com/assets/2198981/10890615/8536c206-815e-11e5-9c51-4a1fa1572daf.png">

birdsarah commented on bokeh/models/tests/test_annotations.py:
```diff
@@ -119,6 +119,37 @@ def test_BoxAnnotation():
+    assert label.x_range_name == 'default'
+    assert label.y_range_name == 'default'
+    assert label.level == 'overlay'
+    yield check_label, label
```
can i suggest that with the addition of `label`, it would be helpful to rename the method `check_label` to, at least, `check_label_text` but perhaps more specifically `check_label_text_properties_defaults`

birdsarah commented on bokehjs/src/coffee/renderer/annotation/label.coffee:
```diff
@@ -0,0 +1,112 @@
+    @text_props = new properties.Text({obj: @model, prefix: 'label_'})
+    @fill_props = new properties.Fill({obj: @model, prefix: 'background_'})
+    @line_props = new properties.Line({obj: @model, prefix: 'border_'})
+    @angle_props = new properties.Angle({obj: @model, attr: 'angle'})
```
sorry for my ignorance on this, why not `angle_`?

canavandl replied:
The initialize method for the Angle property is written differently than the Line/Text/Fill props and doesn't have a 'prefix' attr.

bryevdv replied:
There's some gorpy junk about JS side "properties" that I hope to clear up as part of the new `Document` work in the near future. Short story: line props, etc create several properties at once, and the prefix gives you, e.g., `border_fill_color` but `angle is just a single angle property, called "angle".

birdsarah commented on bokehjs/src/coffee/renderer/annotation/label.coffee:
```diff
@@ -0,0 +1,112 @@
+
+    ctx.restore()
+
+  _calc_dim: (dim, mapper) ->
```
could this have a little js unit test?

birdsarah commented on sphinx/source/docs/user_guide/source_examples/plotting_label.py:
```diff
@@ -0,0 +1,28 @@
+versicolor_label = Label(text='Versicolor', x=5.3, y=1.1, label_text_color='green', **label_props)
+virginica_label = Label(text='Virginica', x=4.4, y=2.2, label_text_color='blue', **label_props)
+
+p.renderers.extend([versicolor_label, virginica_label])
```
I think we stopped manually extending the renderers list and this should be

    p.add_layout(versicolor_label)
    p.add_layout(virginica_label)

bryevdv replied:
That said I think this method name/interface is not great and should be improved (some day)

canavandl commented:
ping @birdsarah 

I added your rec about using the ``add_layout`` plot method. I also refactored the test_annotations.py module to reduce some duplication and make the test method names better. Would you give that a look?

Otherwise, it's ready to merge

birdsarah commented:
Hi @canavandl, I've tried the docs locally while also using the new server docs PR that let you use the local bokeh resources (#3063).

In the docs, the labels plot is appearing above the plot code, but the others in that section have the plot below the code. I think you should be consistent. See v. zoomed out screenshot below.

Also, the labels have a weird dark background on them - I don't think that's intentional. (effect of the new server PR being merged?)

<img width="587" alt="screen shot 2015-11-05 at 7 55 39 pm" src="https://cloud.githubusercontent.com/assets/1796208/10989064/83a10d02-83f7-11e5-9cc3-3b970d55dbc8.png">

birdsarah commented:
If I try and add an empty label `p.add_layout(Label())` my plot displays but without axes and title (presumably because of rendering order) and the console is showing the following error: `Uncaught TypeError: Cannot read property 'indexOf' of null`

This is a "white screen of death" type problem although the screen isn't completely blank because the plot did render. So we should probably provide a warning if you try and instantiate Label like this.

birdsarah commented:
given the labels a full workout here: http://nbviewer.ipython.org/github/birdsarah/bokeh-miscellany/blob/master/tests/annotating%20les%20mis.ipynb

Some things:
- As per the above screenshot, it's annoying that the background box  / border doesn't encapsulate the text.
- angle in radians is working as expected, but not for degrees
- categorical axis are not well supported.

For categorical axis, when I added a box annotation, I was able to do 

```python
p.add_layout(
    BoxAnnotation(
        left='Child1', right='Jondrette',
        top='Child1', bottom='Jondrette',
        fill_color=None, fill_alpha=0,
        line_color='orange', line_width=6,
    )
)
```

but when I added a Label I needed to supply an integer that corresponded to the index of the categorical label in the list:

```python
p.add_layout(
    Label(
        text='Cluster 3', 
        x=reversed_names.index('Jondrette'), 
        y=names.index('Jondrette'),
    )
)
```

what I wanted to do was:

```python
p.add_layout(
    Label(
        text='Cluster 3', 
        x='Jondrette', 
        y='Jondrette',
    )
)
```
but I got the error: `ValueError: expected a value of type Real, got Jondrette of type str`

I'm fine with merging as is, but maybe we should add a new issue to collect these quirks.

damianavila commented:
@canavandl what the status of this one? Btw, it needs some update from master...

bryevdv commented:
@damianavila, @canavandl is occupied with some other tasks for a while, I am going to try and finish this and the other annotation PR up this weekend.

damianavila commented:
Thanks for the info @bryevdv

canavandl commented:
thanks guys!
